File: src/server/events/request.ts
import { ServerConfig, ServerContext } from "../../interfaces";
import { getSchema } from "../utils/schema";
import logger from "../../logger";

import tldjs from "tldjs";
import http from "http";
import url from "url";

import create_server from "./request/create_server";
import handle_request from "./request/handle_request";
import { create_homepage } from "../../pages/home";

export default (context: ServerContext) => {
    const tld = tldjs.fromUserSettings({
        validHosts: [context.config.host],
    });

    return (req: http.IncomingMessage, res: http.ServerResponse) => {
        const subdomain = tld.getSubdomain(req.headers.host);
        if (subdomain) {
            return handle_request(context, req, res, subdomain);
        }

        if (url.parse(req.url, true).query["create"] !== undefined) {
            return create_server(context, req, res);
        }

        return create_homepage(context, req, res);
    };
};

File: src/server/events/request.py
from server_context import ServerContext
from server_config import ServerConfig
from typing import Optional

from tld import get_tld
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse

from create_server import create_server
from handle_request import handle_request
from create_homepage import create_homepage

def get_subdomain(url: str) -> Optional[str]:
    """Get the subdomain from a URL."""
    parsed_url = urlparse(url)
    host = parsed_url.netloc
    tld = get_tld(host, as_object=True)
    return tld.subdomain

def create_request_handler(context: ServerContext):
    """Create a request handler for the server."""
    def request_handler(request: BaseHTTPRequestHandler):
        subdomain = get_subdomain(request.headers.host)

        if subdomain:
            return handle_request(context, request, subdomain)

        if urlparse(request.path, allow_fragments=False).query == "create":
            return create_server(context, request)

        return create_homepage(context, request)

    return request_handler